src/main/java/com/android/volley/toolbox/HurlStack.java

```diff
@@ -289,8 +305,22 @@ private static void addBody(HttpURLConnection connection, Request<?> request, by
             connection.setRequestProperty(
                     HttpHeaderParser.HEADER_CONTENT_TYPE, request.getBodyContentType());
         }
-        DataOutputStream out = new DataOutputStream(connection.getOutputStream());
+        DataOutputStream out = stack.createOutputStream ( request, connection, body.length );
         out.write(body);
-        out.close();
+        out.close ();
+    }
+
+    /**
+     * Overload this method for manipulate request stream (Use 'write (byte[] b)' in subclasse).
+     *
+     * @param request current request.
+     * @param connection current connection of request.
+     * @param length size of stream to write.
+     * @return an DataOutputStream object for write request body.
+     */
+    protected DataOutputStream createOutputStream (Request<?> request, HttpURLConnection connection, 
```

Comment:
We shouldn't require the subclass to create DataOutputStream specifically, as that's an implementation detail. We should just have this method signature require returning an OutputStream, and default to returning connection.getOutputStream(). addBody should still wrap it in a DataOutputStream as before.